src/Modules/OrchardCore.Commerce/Models/ShoppingCartItem.cs

```diff
@@ -124,7 +125,23 @@ public string GetVariantKeyFromAttributes(ISet<string> predefinedAttributeValues
         var price = priceSelectionStrategy.SelectPrice(item.Prices);
         var fullSku = productService.GetOrderFullSku(item, await productService.GetProductAsync(ProductSku));
 
-        var selectedAttributes = Attributes.ToDictionary(key => key.PartName, value => (string)((dynamic)value).PredefinedValue);
+        var selectedTextAttributes = Attributes
+            .Where(attr => attr is TextProductAttributeValue)
+            .ToDictionary(
+                attr => attr.PartName,
+                attr => attr is TextProductAttributeValue textAttribute ? textAttribute.PredefinedValue : null);
+
+        var selectedBooleanAttributes = Attributes
+            .Where(attr => attr is BooleanProductAttributeValue)
+            .ToDictionary(
+                attr => attr.PartName,
+                attr => attr is BooleanProductAttributeValue booleanAttribute ? booleanAttribute.Value.ToString() : null);
+
+        var selectedNumericAttributes = Attributes
+            .Where(attr => attr is NumericProductAttributeValue)
+            .ToDictionary(
+                attr => attr.PartName,
+                attr => attr is NumericProductAttributeValue numericAttribute ? numericAttribute.Value.ToString() : null);
```

Comment:
This too, should be handled by type specific providers.
Also for the boolean and numeric, you can get the same value with `attr.UntypedValue?.ToString()`.